src/main/java/com/tabnine/capabilities/CapabilitiesService.java

```diff
@@ -28,27 +28,71 @@ public void init() {
   }
 
   public boolean isCapabilityEnabled(Capability capability) {
-    if (fetchCapabilitiesFuture == null || !fetchCapabilitiesFuture.isDone()) {
-      return false;
+    synchronized (enabledCapabilities) {
+      return enabledCapabilities.contains(capability);
     }
-    return enabledCapabilities.contains(capability);
   }
 
-  private void scheduleFetchCapabilitiesTask() {
-    fetchCapabilitiesFuture =
-        AppExecutorUtil.getAppScheduledExecutorService()
-            .schedule(this::fetchCapabilities, 2, TimeUnit.SECONDS);
+  private synchronized void scheduleFetchCapabilitiesTask() {
+    if (refreshLoop == null) {
+      refreshLoop = new Thread(() -> this.fetchCapabilitiesLoop());
+      refreshLoop.setDaemon(true);
+      refreshLoop.start();
+    }
+  }
+
+  public static final int INITIAL_DELAY_MS = 2000;
+  public static final int LOOP_INTERVAL_MS = 1000;
+
+  public static final int REFRESH_EVERY_MS = 10 * 60 * 1000; // 10 minutes
+
+  private void fetchCapabilitiesLoop() {
+    Optional<Long> lastRefresh = Optional.empty();
+    Optional<Long> lastPid = Optional.empty();
+
+    try {
```

Comment:
why 2 `try` scopes with the same `catch` implementation? imo only the inner one should exist, so that we'll recover from a single loop iteration's failure

Reply:
The outer ones is to prevent an exception from reaching the user. For every thread we should probably do this.

The inner one is to continue the loop even of something failed.

I agree that the outer one is very unlikely to happen, but it could happen for example if `Thread.sleep` throws ane wexception.